arch/xtensa/core/README-WINDOWS.rst

```diff
@@ -0,0 +1,112 @@
+each call frame that is live in hardware registers, and it will be
+followed by 0, 1 or 2 zero bits that tell you have "big" (how many
+quads of registers) that frame is.
+
+So the CPU executing RETW checks to make sure that the register quad
+being brought into A0-A3 (i.e. the new WINDOWBASE) has a set bit
+indicating it's valid. If it does not, the registers must have been
+spilled and the CPU traps to an exception handler to fill them.
+
+Likewise, the processor can tell if a high register is "owned" by
+another call by seeing if there is a one in WINDOWSTART between that
+register's quad and WINDOWBASE.  If there is, the CPU traps to a spill
+handler to spill one frame.  Note that a frame might be only four
+registers, but it's possible to hit registers 12 out from WINDOWBASE,
+so it's actually possible to trap again when the instruction restarts
+to spill a second quad, and even a third time at maximum.
+
+Finally: note that hardware checks the two bits of WINDOWSTART after
+the frame bit to detect how many quads are represented by the one
+frame.  So there are six separate exception handlers to spill/fill
+1/2/3 quads of registers.
+
+## Software & ABI
+
+The advantage of the scheme above is that it allows the registers to
+be spilled naturally into the stack by using the stack pointers
+embedded in the register file.  But the hardware design assumes and to
+some extent enforces a fairly complicated stack layout to make that
+work:
+
+The spill area for a single frame's A0-A3 registers is not in its own
+stack frame.  It lies in the 16 bytes below its CALLEE's stack
+pointer.  This is so that the callee (and exception handlers invoked
+on its behalf) can see its caller's potentially-spilled stack pointer
+register (A1) on the stack and be able to walk back up on return.
+Other architectures do this too by e.g. pushing the incoming stack
+pointer onto the stack as a standard "frame pointer" defined in the
+platform ABI.  Xtensa wraps this together with the natural spill area
+for register windows.
+
+By convention spill regions always store the lowest numbered register
+in the lowest address.
+
+The spill area for a frame's A4-A11 registers may or may not exist
+depending on whether the call was made with CALL8/CALL12.  It is legal
+to write a function using only A0-A3 and CALL4 calls and ignore higher
+registers.  But if those 0-2 register quads are in use, they appear at
+the top of the stack frame, immediately below the grandparent call's
+A0-A3 spill area.  Because they are at the top of the stack frame,
+they are naturally found not via the frame's own stack pointer but by
+negative offsets from the caller's (!) stack pointer.  This is so that
+A1 can be modified at runtime (e.g. to push/pop data via traditional
+techniques) at runtime without having to have a separate register to
+store the frame pointer -- the caller's stack pointer obviously won't
+change until we return and our frame doesn't exist any more.
```

Comment:
I have a different explanation for this layout. The function knows maximum size of a call it makes, so it can reserve enough space in its own stack frame for the save area for these registers.

Register windows of all callers may be spilled when A1 is modified, so this modification is normally done with the movsp opcode that will take care (by raising the alloca exception) of copying spill area below A1 to a new location if needed.

Reply:
Yeah, this bit could stand to be re-written.  What I think I was trying to discuss here is the way that external code (a debugger or exception handler) finds those registers' spots on the stack.  On other architectures, you just read the frame pointer in e.g. %RBP and offset from that.  Or if that was optimized out the toolchain is expected to give you an elaborate framework for how to extract a base pointer from a current stack pointer and program counter.  But Xtensa doesn't have a frame pointer and doesn't need external metadata, you just roll back to the parent caller in the windows and you can find the current frame without context (albeit in kind of a voodoo way).

And yeah, I never wrote anything about MOVSP/alloca because quite honestly I'm not sure I ever completely understood it. :)

Reply:
> But Xtensa doesn't have a frame pointer

This is not accurate too. Both windowed and call0 ABIs specify frame pointer register (A7 and A15 respectively). But because of the way addressing works in load/store opcodes (only non-negative offsets are available) the frame pointer points not to the top of a frame, but to the bottom. For that reason the frame pointer is less useful for the stack unwinding. But it is till used, e.g. in functions that use alloca.

Reply:
Fair enough, maybe I should just remove the paragraph.  But the point remains: on other architectures a stack frame in the ABI is either unambiguously interpretable without context (e.g. classic x86 debugger can walk the stack of an unknown binary by following nothing but the BP values pushed by the ABI) or requires external context to understand (build with -fomit-frame-pointer and now you really can't do anything without the DWARF info absent a full RE effort).

Xtensa lives in this odd middle space where there is no way for a debugger to reconstruct the stack from a function's own registers, but where it is possible by rolling up the window state.  And that struck me as worth calling out, I guess.

Reply:
> And that struck me as worth calling out, I guess.

Sure, I'm all for it. I'm pointing out that the proposed explanation of the a0-a3 / a4-a11 spill location split wasn't clear to me.